### PR TITLE
Add `TemplateConstraint` and `OrdredConstraint` features (#27706)

### PR DESCRIPTION
**Pull request by jychen630**
# What does this PR do?

According to issue #27706, `TemplateConstraint` is a useful constraint during beam search for constrained decoding. It's desired by the community. In response to this feature request, we implemented the `TemplateConstraint and OrderedConstraint` classes in `src/transformers/generation/beam_constraints.py`. We integrated the constraint workflow with `model.generate()` via corresponding logits processors in `src/transformers/generations/logits_process.py` and test cases in `tests/generation/test_beam_constraints.py`.

Contributors:
Junyao (@jychen630), Raavi (@raavi02) and Pranitha (@NPranitha) 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

Hello!  We noticed @ArthurZucker and @gante 's discussion around template constraints in the issue #27706  - may you please kindly review this a bit? Thanks!

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @gante

**Comment by gante** [comment]
Hi @jychen630 👋 Thank you for opening this PR 🤗 

We're in the process of enabling custom `generate` techniques from the Hub ([PR](https://github.com/huggingface/transformers/pull/36405)). Over time, the goal is to keep the most popular decoding methods in `transformers`, while moving others to the Hub -- this will help our team focus its resources. 

Constrained generation is one of the techniques that will be moved to the Hub. As such, I'm not accepting more features for it at the moment, but I'd like to encourage you to share your expanded functionality as a Hub generation method! The PR I linked above should be merged this week :D
